models/gif_psc_exp.cpp

```diff
@@ -0,0 +1,469 @@
+
+  double_t tau_m = P_.c_m_ / P_.g_L_;
+
+  // these are determined according to a numeric stability criterion
+  V_.P21ex_ = propagator_32( P_.tau_ex_, tau_m, P_.c_m_, V_.h_ );
+  V_.P21in_ = propagator_32( P_.tau_in_, tau_m, P_.c_m_, V_.h_ );
+
+
+  V_.P33_ = std::exp( -V_.h_ / tau_m );
+  V_.P30_ = 1 / P_.c_m_ * ( 1 - V_.P33_ ) * tau_m;
+  V_.P31_ = ( 1 - V_.P33_ );
+
+  V_.RefractoryCounts_ = Time( Time::ms( P_.t_ref_ ) ).get_steps();
+  assert( V_.RefractoryCounts_ >= 0 ); // since t_ref_ >= 0, this can only fail in error
+
+  // initializing internal state
+  if ( !S_.initialized_ )
+  {
+    for ( uint_t i = 0; i < P_.tau_sfa_.size(); i++ )
+    {
+      V_.Q33_.push_back( std::exp( -V_.h_ / P_.tau_sfa_[ i ] ) );
+      S_.q_sfa_elems_.push_back( 0.0 );
+    }
+
+    for ( uint_t i = 0; i < P_.tau_stc_.size(); i++ )
+    {
+      V_.Q44_.push_back( std::exp( -V_.h_ / P_.tau_stc_[ i ] ) );
+      S_.q_stc_elems_.push_back( 0.0 );
+    }
+
+    S_.initialized_ = true;
+  }
+}
+
+/* ----------------------------------------------------------------
+ * Update and spike handling functions
+ */
+
+void
+nest::gif_psc_exp::update( Time const& origin, const long_t from, const long_t to )
+{
+
+  assert( to >= 0 && ( delay ) from < kernel().connection_builder_manager.get_min_delay() );
+  assert( from < to );
+
+  double_t q_temp_;
+
+  for ( long_t lag = from; lag < to; ++lag )
+  {
+
+    q_temp_ = 0;
+    for ( uint_t i = 0; i < S_.q_stc_elems_.size(); i++ )
+    {
+
+      q_temp_ += S_.q_stc_elems_[ i ];
+
+      S_.q_stc_elems_[ i ] = V_.Q44_[ i ] * S_.q_stc_elems_[ i ];
+    }
+
+    S_.stc_ = q_temp_;
+
+    q_temp_ = 0;
+    for ( uint_t i = 0; i < S_.q_sfa_elems_.size(); i++ )
+    {
+
+      q_temp_ += S_.q_sfa_elems_[ i ];
+
+      S_.q_sfa_elems_[ i ] = V_.Q33_[ i ] * S_.q_sfa_elems_[ i ];
+    }
+
+    S_.q_ = q_temp_ + P_.v_t_star_;
+
+    ulong_t n_spikes = 0;
+
+    if ( S_.r_ref_ == 0 ) // neuron not refractory, so evolve V
+    {
+
+      if ( S_.add_stc_sfa_ == true )
+      {
+
+        S_.add_stc_sfa_ = false;
+
+
+        q_temp_ = 0;
+
+        for ( uint_t i = 0; i < S_.q_stc_elems_.size(); i++ )
+        {
+          S_.q_stc_elems_[ i ] += P_.q_stc_[ i ];
+
+          q_temp_ += P_.q_stc_[ i ];
+        }
+
+        S_.stc_ += q_temp_;
+
+
+        q_temp_ = 0;
+
+        for ( uint_t i = 0; i < S_.q_sfa_elems_.size(); i++ )
+        {
+
+          S_.q_sfa_elems_[ i ] += P_.q_sfa_[ i ];
+
+          q_temp_ += P_.q_sfa_[ i ];
+        }
+
+        S_.q_ += q_temp_;
+      }
+
+      S_.y3_ = V_.P30_ * ( S_.y0_ + P_.I_e_ - S_.stc_ ) + V_.P33_ * S_.y3_ + V_.P31_ * P_.E_L_
+        + S_.i_syn_ex_ * V_.P21ex_ + S_.i_syn_in_ * V_.P21in_;
+
+      double_t lambda = P_.lambda0_ * std::exp( ( S_.y3_ - S_.q_ ) / P_.delta_u_ );
+
+      if ( lambda > 0.0 )
+      {
+
+        // Draw random number and compare to prob to have a spike
+        if ( V_.rng_->drand() <= -numerics::expm1( -lambda * ( V_.h_ / 1000.0 ) ) )
```

Comment:
This should be `<`, not `<=`, since `drand()` returns numbers from `[0, 1)´, not `(0, 1]`. You can avoid repeated division by 1000 by rescaling lambda0 to 1/ms when the parameter is set. You should also comment why the `expm1()` function appears here. Most people would expect just `lambda*h`, but you approximate the integral in Eq 3 of the PLoS paper as `lambda * h` and then use the more exact expression `1 - exp(-lambda*h)`, don't you?